os/linux/linux_frontend.cpp

```diff
@@ -0,0 +1,401 @@
+                            iahwc_callback_data_t data,
+                            iahwc_function_ptr_t hook) {
+  switch (description) {
+    case IAHWC_CALLBACK_VSYNC: {
+      if (display_id >= displays_.size())
+        return IAHWC_ERROR_BAD_DISPLAY;
+      IAHWCDisplay* display = displays_.at(display_id);
+      return display->RegisterVsyncCallback(data, hook);
+    }
+    default:
+      return IAHWC_ERROR_BAD_PARAMETER;
+  }
+}
+
+IAHWC::IAHWCDisplay::IAHWCDisplay() : native_display_(NULL) {
+}
+
+int IAHWC::IAHWCDisplay::Init(hwcomposer::NativeDisplay* display) {
+  native_display_ = display;
+  native_display_->InitializeLayerHashGenerator(4);
+}
+
+int IAHWC::IAHWCDisplay::GetDisplayInfo(uint32_t config, int attribute,
+                                        int32_t* value) {
+  hwcomposer::HWCDisplayAttribute attrib =
+      static_cast<hwcomposer::HWCDisplayAttribute>(attribute);
+
+  bool ret = native_display_->GetDisplayAttribute(config, attrib, value);
+
+  if (!ret)
+    return IAHWC_ERROR_NO_RESOURCES;
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::GetDisplayName(uint32_t* size, char* name) {
+  bool ret = native_display_->GetDisplayName(size, name);
+
+  if (!ret)
+    return IAHWC_ERROR_NO_RESOURCES;
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::GetDisplayConfigs(uint32_t* num_configs,
+                                           uint32_t* configs) {
+  bool ret = native_display_->GetDisplayConfigs(num_configs, configs);
+
+  if (!ret)
+    return IAHWC_ERROR_NO_RESOURCES;
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::SetDisplayGamma(float r, float b, float g) {
+  native_display_->SetGamma(r, g, b);
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::SetDisplayConfig(uint32_t config) {
+  bool ret = native_display_->SetActiveConfig(config);
+
+  if (!ret)
+    return IAHWC_ERROR_NO_RESOURCES;
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::GetDisplayConfig(uint32_t* config) {
+  bool ret = native_display_->GetActiveConfig(config);
+
+  if (!ret)
+    return IAHWC_ERROR_NO_RESOURCES;
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::ClearAllLayers() {
+  layers_.clear();
+  native_display_->ResetLayerHashGenerator();
+
+  return IAHWC_ERROR_NONE;
+}
+int IAHWC::IAHWCDisplay::PresentDisplay(int32_t* release_fd) {
+  std::vector<hwcomposer::HwcLayer*> layers;
+  hwcomposer::HwcLayer* cursor_layer = NULL;
+
+  for (auto & [ first, second ] : layers_) {
+    if (second.GetLayerUsage() == IAHWC_LAYER_USAGE_CURSOR)
+      cursor_layer = second.GetLayer();
+    else
+      layers.emplace_back(second.GetLayer());
+  }
+
+  if (cursor_layer)
+    layers.insert(layers.begin(), cursor_layer);
+
+  native_display_->Present(layers, release_fd);
+
+  return IAHWC_ERROR_NONE;
+}
+
+int IAHWC::IAHWCDisplay::CreateLayer(uint32_t* layer_handle) {
+  *layer_handle = native_display_->AcquireId();
+  layers_.emplace(*layer_handle, IAHWCLayer());
```

Comment:
Where is the counterpart destroy layer call ?, I dont see us giving back this layer id anywhere.

Reply:
Yeah, I didn't put that in as for now I am just destroying all the layers and creating them every iteration. Will add api for destroy layer.

Reply:
Merged this for now. Add the new api which will release the layer handle.

Reply:
sure :)